machine.go

```diff
@@ -162,116 +158,89 @@ func NewMachine(cfg Config, opts ...Opt) (*Machine, error) {
-
-	if err = m.createMachine(ctx); err != nil {
-		m.stopVMM()
-		return errCh, err
-	}
-	m.logger.Debug("createMachine returned")
-
-	if err = m.createBootSource(ctx, m.cfg.KernelImagePath, m.cfg.KernelArgs); err != nil {
+// Start will iterate through the handler list and call each
+// handler. If an error occurred during handler execution, that
+// error will be returned. If the handlers succeed, then this
+// will start the VMM instance.
+func (m *Machine) Start(ctx context.Context) error {
+	m.logger.Debug("Called Machine.Start()")
+	if err := m.Handlers.Run(ctx, m); err != nil {
 		m.stopVMM()
-		return errCh, err
+		return err
 	}
-	m.logger.Debug("createBootSource returned")
 
-	if err = m.attachDrive(ctx, m.cfg.RootDrive, 1, true); err != nil {
-		m.stopVMM()
-		return errCh, err
-	}
-	m.logger.Debug("Root drive attachment complete")
+	return m.StartInstance(ctx)
+}
 
-	for id, dev := range m.cfg.AdditionalDrives {
-		// id must be increased by 2 because firecracker uses 1-indexed arrays and the root drive occupies position 1.
-		err = m.attachDrive(ctx, dev, id+2, false)
-		if err != nil {
-			m.logger.Errorf("While attaching secondary drive %s, got error %s", dev.HostPath, err)
+func (m *Machine) addVsocks(ctx context.Context, vsocks ...VsockDevice) error {
+	for _, dev := range m.cfg.VsockDevices {
+		if err := m.addVsock(ctx, dev); err != nil {
 			m.stopVMM()
-			return errCh, err
+			return err
 		}
-		m.logger.Debugf("attachDrive returned for %s", dev.HostPath)
 	}
-	for id, iface := range m.cfg.NetworkInterfaces {
-		err = m.createNetworkInterface(ctx, iface, id+1)
-		if err != nil {
-			m.stopVMM()
-			return errCh, err
+	return nil
+}
+
+func (m *Machine) createNetworkInterfaces(ctx context.Context, ifaces ...NetworkInterface) error {
+	for id, iface := range ifaces {
+		if err := m.createNetworkInterface(ctx, iface, id+1); err != nil {
+			return err
 		}
 		m.logger.Debugf("createNetworkInterface returned for %s", iface.HostDevName)
 	}
-	for _, dev := range m.cfg.VsockDevices {
-		err = m.addVsock(ctx, dev)
-		if err != nil {
-			m.stopVMM()
-			return errCh, err
+
+	return nil
+}
+
+func (m *Machine) attachDrives(ctx context.Context, rootIndex int, drives ...BlockDevice) error {
+	for id, dev := range drives {
+		if err := m.attachDrive(ctx, dev, id+1, id == rootIndex); err != nil {
+			m.logger.Errorf("While attaching secondary drive %s, got error %s", dev.HostPath, err)
+			return err
 		}
+		m.logger.Debugf("attachDrive returned for %s", dev.HostPath)
 	}
 
-	m.logger.Debugf("returning from Machine.Init(), RootDrive=%s", m.cfg.RootDrive.HostPath)
-	return errCh, nil
+	return nil
 }
 
 // startVMM starts the firecracker vmm process and configures logging.
-func (m *Machine) startVMM(ctx context.Context) (<-chan error, error) {
+func (m *Machine) startVMM(ctx context.Context) error {
 	m.logger.Printf("Called startVMM(), setting up a VMM on %s", m.cfg.SocketPath)
 
-	exitCh := make(chan error)
+	m.ErrCh = make(chan error)
```

Comment:
I don't think I really like `m.ErrCh` being public.  This function uses the channel for internal signalling before returning, and I don't think it wise for that to be exposed while the function is executing.  Having it in the return meant that it was inaccessible until the function returned, but moving it into the struct and making it public loses that.  I think I'd prefer it to just be returned with the function, but if you think that's not possible, an alternative is not assigning it to the public field in the struct until the end of this function, after its internal dependencies are done trying to read from the channel (i.e., after the call to `waitForSocket`).